hera_cal/abscal_funcs.py

```diff
@@ -996,111 +986,73 @@ def array_axis_to_data_key(data, array_index, array_keys, key_index=-1, copy_dic
+    kernel += tuple(np.ones((data.ndim - len(kernel)), np.int))
     if medfilt:
-        vis_smooth = signal.medfilt(np.real(vis), kernel_size=kernel) + 1j * signal.medfilt(np.imag(vis), kernel_size=kernel)
-    else:
-        vis_smooth = vis
+        data = signal.medfilt(np.real(data), kernel_size=kernel) + 1j * signal.medfilt(np.imag(data), kernel_size=kernel)
 
     # construct window
-    win = np.moveaxis(np.repeat(np.zeros(Nfreqs)[np.newaxis], Ntimes, axis=0), 0, time_ax)
-
-    if edge_cut > 0:
-        assert 2 * edge_cut < Nfreqs - 1, "edge_cut cannot be >= Nfreqs/2 - 1"
-        win_slice = slice(edge_cut, Nfreqs - edge_cut)
-        win_Nfreqs = Nfreqs - 2 * edge_cut
-    else:
-        win_slice = slice(None)
-        win_Nfreqs = Nfreqs
-
-    if window is None:
-        # this is a top-hat window
-        win[:, win_slice] = signal.windows.tukey(win_Nfreqs, 0.0)
+    win = np.zeros_like(data, dtype=np.float)
+    assert 2 * edge_cut < Nfreqs - 1, "edge_cut cannot be >= Nfreqs/2 - 1"
+    if window is None:  # tophat window
+        win[:, edge_cut:(Nfreqs - edge_cut)] = 1.0
     elif window == 'blackmanharris':
-        win[:, win_slice] = signal.windows.blackmanharris(win_Nfreqs)
+        win[:, edge_cut:(Nfreqs - edge_cut)] = signal.windows.blackmanharris(Nfreqs - 2 * edge_cut)
     elif window == 'hann':
-        win[:, win_slice] = signal.windows.hann(win_Nfreqs)
+        win[:, edge_cut:(Nfreqs - edge_cut)] = signal.windows.hann(Nfreqs - 2 * edge_cut)
     else:
         raise ValueError("didn't recognize window {} from ['blackmanharris', 'hann', None]".format(window))
 
-    # multiply wgts
-    win *= wgts
-
-    # fft w/ window
-    vfft = np.fft.fft(vis_smooth * win, axis=freq_ax)
-
-    # get argmax of abs
-    amp = np.abs(vfft)
-    argmax = np.moveaxis(np.argmax(amp, axis=freq_ax)[np.newaxis], 0, freq_ax)
-
-    # get delays
-    fftfreqs = np.fft.fftfreq(Nfreqs, 1)
-    dfreq = np.median(np.diff(fftfreqs))
-    dlys = fftfreqs[argmax]
-
-    # get peak shifts, and add to dlys
-    def get_peak(amp, max_ind):
-        Nchan = len(amp)
-        y = np.concatenate([amp, amp, amp])
-        max_ind += Nchan
-        y = y[max_ind - 1:max_ind + 2]
-        r = np.abs(np.diff(y))
-        r = r[0] / r[1]
-        peak = 0.5 * (r - 1) / (r + 1)
-        return peak
-
-    peak_shifts = np.array([get_peak(np.take(amp, i, axis=time_ax), np.take(argmax, i, axis=time_ax)[0]) for i in range(Ntimes)])
-    dlys += np.moveaxis(peak_shifts.reshape(-1, 1) * dfreq, 0, time_ax)
+    # fft w/ window and find argmax
+    vfft = np.fft.fft(data * win * wgts, axis=1)
+    amp = np.abs(np.fft.fftshift(vfft))
+    argmaxes = np.argmax(amp, axis=1)
+    fftfreqs = np.fft.fftshift(np.fft.fftfreq(Nfreqs, df))
+    dlys, peak_shifts = np.zeros((Ntimes, 1)), np.zeros((Ntimes, 1))
+    # use parabolic peak interpolation: https://ccrma.stanford.edu/~jos/sasp/Quadratic_Interpolation_Spectral_Peaks.html
```

Comment:
A few comments on the new peak finder / shifter:
1) Is it faster or slower than the one before? The reason I adopted the previous approach was that it was about as fast as way as I could think of doing this, and its an analytic result. 2) It seems like what you've implemented is actually quite similar (`0.5 * something / something`), and 3) does it give the same or nearly similar result as the old function? I'm happy to go with this new implementation if the speed isn't significantly slower and its getting similar if not the same results as before

Reply:
Timing on a (4354, 1024) complex array:

New version takes 365 ms, old version takes 332 ms. The results are identical to numerical precision, so I think we're just writing the same algorithm differently.

And for the record, the old version using the default kwargs takes 2m 57s (the 332 ms is without the medfilt or the phase finding).